lib/apiservers/engine/backends/container.go

```diff
@@ -345,137 +334,288 @@ func (c *ContainerBackend) ContainerExecResize(eid string, height, width int) er
-		if !attach {
-			op.Debugf("Detached mode. Returning early.")
-			return nil
-		}
+	op.Debugf("Exec start of %s.%s", eid, id)
 
-		EventService().Log(containerAttachEvent, eventtypes.ContainerEventType, actor)
-		ca := &backend.ContainerAttachConfig{
-			UseStdin:  ec.OpenStdin,
-			UseStdout: ec.OpenStdout,
-			UseStderr: ec.OpenStderr,
-		}
+	var ec *models.TaskInspectResponse
+	operation := func() error {
+		var err error
+		ec, err = c.taskStartHelper(op, id, eid, name)
+		return err
+	}
 
-		// set UseStderr to false for Tty case as we merge stdout and stderr
-		if ec.Tty {
-			ca.UseStderr = false
-		}
+	if err := retry.DoWithConfig(operation, engerr.IsConflictError, backoffConf); err != nil {
+		op.Errorf("Failed to start Exec task for container(%s) due to error (%s)", id, err)
+		return err
+	}
 
-		ac := &proxy.AttachConfig{
-			ID: eid,
-			ContainerAttachConfig: ca,
-			UseTty:                ec.Tty,
-			CloseStdin:            true,
-		}
+	attach := ec.OpenStdin || ec.OpenStdout || ec.OpenStderr
 
-		err = c.streamProxy.AttachStreams(ctx, ac, stdin, stdout, stderr)
-		if err != nil {
-			if _, ok := err.(engerr.DetachError); ok {
-				op.Infof("Detach detected, tearing down connection")
+	// we need to be able to cancel it
+	taskOp, cancel := trace.WithCancel(&op, "exec task wait on %s", eid)
+	defer cancel()
 
-				// QUESTION: why are we returning DetachError? It doesn't seem like an error
-				// fire detach event
-				EventService().Log(containerDetachEvent, eventtypes.ContainerEventType, actor)
+	attachResult := make(chan error, 1)
+	startResult := make(chan error, 1)
 
-				// DON'T UNBIND FOR NOW, UNTIL/UNLESS REFERENCE COUNTING IS IN PLACE
-				// This avoids cutting the communication channel for other sessions connected to this
-				// container
+	if attach {
+		go func() {
+			attachResult <- c.attachHelper(taskOp, ec, stdin, stdout, stderr, false)
+			close(attachResult)
+		}()
+	}
 
-				// FIXME: call UnbindInteraction/Commit
+	go func() {
+		targetState := map[string]bool{constants.TaskRunningState: true, constants.TaskStoppedState: true}
+		waitState := map[string]bool{constants.TaskCreatedState: true, constants.TaskUnknownState: true}
+		_, err := c.taskStateWaitHelper(taskOp, id, eid, name, targetState, waitState)
+		if err != nil {
+			op.Errorf("Wait for exec start on %s.%s: %s", eid, id, err)
+			startResult <- err
+		}
+
+		close(startResult)
+	}()
+
+	// wait for either wait to succeed, fail, or for stream connection to error out
+	select {
+	case err := <-startResult:
+		if err != nil {
+			op.Errorf("Task wait returned error: %s", err)
+			// we can't return a proper error as we close the streams as soon as AttachStreams returns so we mimic Docker and write to stdout directly
+			// https://github.com/docker/docker/blob/a039ca9affe5fa40c4e029d7aae399b26d433fe9/api/server/router/container/exec.go#L114
+			if stdout != nil {
+				stdout.Write([]byte(err.Error() + "\r\n"))
```

Comment:
Check if this is redundant in the client version we support. I'm seeing a double echo of the error during `-it` runs for executables that do not exist.